raft/raft.cpp

```diff
@@ -617,39 +727,38 @@ raft::handle_request_append_entries_response(const bzn::message& msg, std::share
         return;
     }
 
-    // copy over match_log_index
-    std::vector<uint32_t> match_indexes;
-    for(const auto& match_index : this->peer_match_index)
-    {
-        match_indexes.push_back(match_index.second);
-    }
-    std::sort(match_indexes.begin(), match_indexes.end());
-    size_t consensus_commit_index = match_indexes[uint32_t(std::ceil(match_indexes.size()/2.0))];
-
-    while (this->commit_index < consensus_commit_index)
+    while (this->commit_index < this->last_majority_replicated_log_index())
     {
-        this->perform_commit(this->commit_index, this->log_entries[this->commit_index]);
+        this->perform_commit(this->commit_index, this->raft_log->entry_at(this->commit_index));
     }
 }
 
 
 bool
-raft::append_log(const bzn::message& msg)
+raft::append_log_unsafe(const bzn::message& msg, const bzn::log_entry_type entry_type)
 {
-    std::lock_guard<std::mutex> lock(this->raft_lock);
-
     if (this->current_state != bzn::raft_state::leader)
     {
-        LOG(warning) << "not the leader can't append log_entries!";
+        LOG(warning) << "not the leader, can't append log_entries!";
         return false;
     }
 
-    this->log_entries.emplace_back(log_entry{bzn::log_entry_type::log_entry, ++this->last_log_index, this->current_term, msg});
+    LOG(debug) << "Appending " << log_entry_type_to_string(entry_type) << " to my log: " << msg.toStyledString();
+
+    this->raft_log->leader_append_entry(log_entry{entry_type, uint32_t(this->raft_log->size()), this->current_term, msg});
 
     return true;
 }
 
 
+bool
+raft::append_log(const bzn::message& msg, const bzn::log_entry_type entry_type)
+{
+    std::lock_guard<std::mutex> lock(this->raft_lock);
+    return this->append_log_unsafe(msg, entry_type);
```

Comment:
you are locking to I guess unsafe is now safe?

Reply:
We pulled the code that is in append_log_unsafe from this function, so we could reuse it in it's unsafe form without blocking.